File: src/main/java/com/github/nigelzor/hogs/Direction.java
package com.github.nigelzor.hogs;

import java.util.EnumSet;
import java.util.Set;

import com.google.common.base.Function;

public enum Direction implements Function<Index, Index> {
	NORTH(-1, 0),
	EAST(0, 1),
	SOUTH(1, 0),
	WEST(0, -1);

	private final int row;
	private final int col;

	private Direction(int row, int col) {
		this.row = row;
		this.col = col;
	}

	public Direction rotate(Rotation rotation) {
		return Direction.values()[(ordinal() + rotation.ordinal()) % 4];
	}

	public static Set<Direction> rotate(Set<Direction> directions, Rotation rotation) {
		EnumSet<Direction> result = EnumSet.noneOf(Direction.class);
		for (Direction direction : directions) {
			result.add(direction.rotate(rotation));
		}
		return result;
	}

	@Override
	public Index apply(Index input) {
		return new Index(input.getRow() + row, input.getCol() + col);
	}

}

File: src/main/kotlin/com/github/nigelzor/hogs/Direction.kt
package com.github.nigelzor.hogs

enum class Direction(val row: Int, val col: Int) {
	NORTH(-1, 0),
	EAST(0, 1),
	SOUTH(1, 0),
	WEST(0, -1);

	val bits = 1 shl ordinal

	fun rotate(rotation: Rotation): Direction {
		return Direction.values()[(ordinal + rotation.ordinal) % 4]
	}

	fun apply(input: Index): Index {
		return Index(input.row + row, input.col + col)
	}
}
